### ISSUE TITLE
Timer()

### ISSUE THREAD

**Issue by GralfR** [hossimo/GMA3Plugins#36]
Some more info on the Timer() Function:

If the parameter maxCount is zero, then the timer runs forever. If the parameter is not zero, then the timer runs the given number of loops. Compared to Timer inside LUA on MA2.

The parameter "cleanup" is the name of a function, that is called, when the timer is stopped. You can use this to run a certain task once the plugin is stopped. I use this to restart the plugin when it should run forever (maxCount=0) and was stopped by user.
See issue #35 for an example.

**Comment by hossimo** [comment]
Hey there thanks for the note, I'll add this later.

Just a guess here but I believe there is another parameter that you can pass to Timer(), it seems to give it some context when stopping. I'm not 100% sure since the internal documentation is not clear and I can't find a proper example except one that uses signals that I can't dig into right now.

The function signature seems to be the following:

``` lua
Timer(
        function : name,
        number : delaytime,
        number : max_count,
        [function : cleanup],
        [light_userdata : context]
) -> nothing
```


If I pass anything to the context parameter it just does not call the cleanup function. so my guess is this stop call happens outside of the context of the user's runtime so it does not have access to the CurrnetUser() for like you did in #35 thus it gets a nil value causing the error.

I tried passing the handle:
```  lua
local handle = select(4,...);
Timer(running, 1, 0, stop, handle)
```

as well as the display_handle:
``` lua
local function start(display_handle)
    Timer(running, 1, 0, stop, display_handle)
end
```

I even tried passing `self`:

``` lua
    Timer(running, 1, 0, stop, self)
```

but each just causes the cleanup function not to run, I wonder what this value should be?

anyway, testing for nil is a workaround for the moment.

``` lua
local function stop()
	Echo("Stopping")
	if CurrentUser() == nil then
		Echo("Timer Stopped")
	else
		Printf("timer stopped by "..CurrentUser().Name)
	end
end
```

**Comment by GralfR** [comment]
Thanks for testing and the effort with examples. Testing the CurrentUser() for nil is just a workaround to not let the plugin crash on cleanup, but it's not an option for my initial goal: testing, if the user has the right to stop the plugin and the timer. This was easy in MA2, but a big problem on MA3, since a guest-user with playback-rights may stop a plugin but is not allowed to start it (bug already reported to MA). My goal was to check not only the user-name but the user-rights the current user has, when the plugin was stopped and then for example try to restart the plugin or do other things. This was no problem on MA2.

Concerning the last argument I have no clue at all. I did not understand the whole object-structur on MA3 LUA yet. The last argument did not exist in MA2 LUA. So I can't explain, what it might be for. Maybe it's not yet implemented because it's optional. Or the hint "context" might give a clue: maybe it's the environment, the timer should run, or it's the context in which the timer was stopped. Maybe it's not an argument passed to the timer but a placeholder for a feedback/return-value of the timer?
I just started the transition from MA2 LUA to MA3 and it's very hard to learn new stuff, that is not well documented. What is the object-structure of MA3? What's about the "signals" or "hooks"? Are these trigger/interrupts? I never got deep into LUA. Just the few things is needed for MA2. Tried to dig a bit deeper into LUA but was stuck soon because MA2 did not have certain libraries included. Maybe this is better on MA3, but nobody knows because of the lack of good introduction, what MA3 means for LUA. I can get LUA tutorials everywhere, but they don't know about MA3-objects/functions. So it stays a nightmare to reverse-engineer and spend a lot of time for trial and error. Thanks for Your effort in this! Hope You'll find enough time to keep this up and maybe write a MA3-LUA-Beginners-Tutorial for anybody that knows LUA, but needs a MA3 LUA jumpstart.
